### ISSUE TITLE
Some confusion about the ip_layer.py file

### ISSUE THREAD

**Issue by zcx-language** [axelBarroso/Key.Net#8]
Hi, Really appreciate for sharing the code and I have read your code carefully for days. 
And I am confused about the code lines below

`weights = tf.nn.max_pool(tf.stop_gradient(scores), [1, window_size, window_size, 1], strides=[1, window_size, window_size, 1], padding='VALID')

max_pool_unpool = tf.nn.conv2d_transpose(weights, kernels['upsample_filter_np_'+str(window_size)],
                                              output_shape=[shape_scores[0], shape_scores[1], shape_scores[2], 1],
                                              strides=[1, window_size, window_size, 1])

exp_map_1 = tf.add(tf.pow(exponential_value, tf.div(scores, max_pool_unpool+1e-6)), -1*(1.-1e-6))
`
in ip_layer function. 

my question is max_pool_unpool may have many zero value because of conv_transpose, why did you divide
scores by max_poo_unpool?,  and this may cause many inf value because of e^(1e6)?
Can you explain the  calculation of  the max_pool_unpool and tf.div(scores, max_pool_unpool) ?

Really Thanks!
for example:
`when 

score = tf.Variable([[[[6.], [2.], [3.], [4.]], [[5.], [6.], [7.], [8.]], [[9], [10], [11], [12]], [[13], [14], [16], [17]]]])
window_size = 2

then 

max_pool_unpool = [[[[6][0][8][0]][[0][0][0][0]][[14][0][17][0]][[0][0][0][0]]]]
exp_map_1 = [[[[1.71828222][inf][0.454992473][inf]][[inf][inf][inf][inf]][[0.9019081][inf][0.909916][inf]][[inf][inf][inf][inf]]]]
`

`def ip_layer(scores, window_size, kernels):

    exponential_value = np.e

    shape_scores = tf.shape(scores)

    weights = tf.nn.max_pool(tf.stop_gradient(scores), [1, window_size, window_size, 1], strides=[1, window_size, window_size, 1], padding='VALID')

    max_pool_unpool = tf.nn.conv2d_transpose(weights, kernels['upsample_filter_np_'+str(window_size)],
                                              output_shape=[shape_scores[0], shape_scores[1], shape_scores[2], 1],
                                              strides=[1, window_size, window_size, 1])

    exp_map_1 = tf.add(tf.pow(exponential_value, tf.div(scores, max_pool_unpool+1e-6)), -1*(1.-1e-6))

    sum_exp_map_1 = tf.nn.conv2d(exp_map_1, kernels['ones_kernel_'+str(window_size)], [1, window_size, window_size, 1], padding='VALID')

    indexes_map = tf.nn.conv2d(exp_map_1, kernels['indexes_kernel_' + str(window_size)], [1, window_size, window_size, 1], padding='VALID')

    indexes_map = tf.divide(indexes_map, tf.add(sum_exp_map_1, 1e-6))

    max_weights = tf.reduce_max(weights, axis=[1, 2, 3], keepdims=True)

    norm_weights = tf.divide(weights, max_weights + 1e-6)

    return indexes_map, [weights, norm_weights]
`

**Comment by zcx-language** [comment]
OK, I understood. window_size won't be equal to 2, and window_size = 8, 16... 
it just let the score map divided by its local maximum within a window , is that right?

**Comment by axelBarroso** [comment]
Hi, 

Thanks for posting your question! 

Yes, that is right. The idea of the IP layer is to get the maximum within a window. Such maximum is used to normalize the windows before the spatial soft-max. The idea is to get the contribution of each pixel within the window and compute its index position. 

1e-6 values are used for regularization since it could happen that the maximum in a region is 0.0, and therefore it would cause "inf" values.
